flamingpy/noise/cv.py

```diff
@@ -0,0 +1,623 @@
+
+    Args:
+        code (SurfaceCode or EGraph): the code object (so that code.graph is
+            an EGraph), or an EGraph directly, to which the noise is applied.
+        delta (float): the quadrature blurring parameter, related to the
+            squeezing of the GKP states and the momentum-quadrature variance of
+            the p-squeezed states.
+        states (dict, optional): the dictionary of all non-GKP states and their
+            indices, of the form {'state': []}. By default, all states are
+            treated as GKP states.
+        p_swap (float, optional): if supplied, the probability of a node being
+            a p-squeezed state. Overrides the indices given in state.
+        sampling_order (str, optional): the scheme for conducting sampling
+            for the homodyne measurements. Options are "initial" and "two-step".
+        rng (numpy.random.Generator, optional): a random number generator
+            following the NumPy API. It can be seeded for reproducibility.
+            By default, numpy.random.default_rng is used without a fixed seed.
+
+    Attributes:
+        egraph (EGraph): the underlying graph representation.
+        to_points (dict): the index-to-coordinate dictionary, taken from egraph.
+        N (int): the number of qubits in the lattice.
+        delta (float): the delta from the Args above.
+        p_swap (float): the swap-out probability from the Args above.
+        _adj (sp.sparse.csr_matrix): the adjacency matrix of egraph.
+        _states (dict): states along with their indices.
+        _sampling_order (str): sampling order from above.
+        _perfect_inds (list or NoneType): the indices of qubits to treat as
+            ideal.
+    """
+
+    def __init__(self, code, *, delta, **kwargs):
+        # Point to the EGraph, if supplied directly, or to code.egraph if not.
+        if isinstance(code, EGraph):
+            self.egraph = code
+        else:
+            self.code = code
+            self.egraph = code.graph
+
+        # Set some egraph properties to attributes of self.
+        # (Note: self._adj also generates indices in the EGraph)
+        self._adj = self.egraph.adj_generator(sparse=True)
+        self._N = len(self.egraph)
+        self._to_points = self.egraph.to_points
+
+        # Get indices of nodes we wish to designate as ideal.
+        perfect_points = self.egraph.graph.get("perfect_points")
+        if perfect_points is None:
+            self._perfect_inds = None
+        else:
+            self._perfect_inds = [self.egraph.to_indices[point] for point in perfect_points]
+
+        # Set some noise model parameters.
+        self.delta = delta
+        self.p_swap = kwargs.get("p_swap")
+        self.states = kwargs.get("states") or {"p": np.empty(0, dtype=int)}
+        self.sampling_order = kwargs.get("sampling_order")
+
+        # Assign state labels.
+        self.populate_states()
```

Comment:
This step should not be done when creating a `CvLayer`. Anything that require randomness and that change for each Monte Carlo iteration should be done in `apply_noise`. This way, we can build the `CvLayer` once for the whole simulation. This is not really for efficiency, but for clarity and consistency.

Reply:
Yes, good catch -- I made this change in a recent commit so that the user did not have to explicitly populate states when instantiating a CVLayer. I will revert this.